src/Agent.Plugins/PipelineArtifact/FileShareProvider.cs

```diff
@@ -0,0 +1,145 @@
+
+        public async Task PublishArtifactAsync(string sourcePath, string destPath, int parallelCount, CancellationToken cancellationToken)
+        {
+            await PublishArtifactUsingRobocopyAsync(this.context, new HostContext(this.hostType), sourcePath, destPath, cancellationToken);
+        }
+
+        private async Task CopyFileShareAsync(string downloadRootPath, string destPath, IEnumerable<string> minimatchPatterns, CancellationToken cancellationToken)
+        {
+            IEnumerable<Func<string, bool>> minimatcherFuncs = MinimatchHelper.GetMinimatchFuncs(minimatchPatterns, this.tracer);
+            await DownloadFileShareArtifactAsync(downloadRootPath, destPath, defaultParallelCount, cancellationToken, minimatcherFuncs);
+        }
+        
+        private async Task PublishArtifactUsingRobocopyAsync(AgentTaskPluginExecutionContext executionContext, IHostContext hostContext, string dropLocation, string downloadFolderPath, CancellationToken cancellationToken)
+        {
+            executionContext.Output(StringUtil.Loc("PublishingArtifactUsingRobocopy"));
+            using (var processInvoker = hostContext.CreateService<IProcessInvoker>())
+            {
+                // Save STDOUT from worker, worker will use STDOUT report unhandle exception.
+                processInvoker.OutputDataReceived += delegate (object sender, ProcessDataReceivedEventArgs stdout)
+                {
+                    if (!string.IsNullOrEmpty(stdout.Data))
+                    {
+                        executionContext.Output(stdout.Data);
+                    }
+                };
+
+                // Save STDERR from worker, worker will use STDERR on crash.
+                processInvoker.ErrorDataReceived += delegate (object sender, ProcessDataReceivedEventArgs stderr)
+                {
+                    if (!string.IsNullOrEmpty(stderr.Data))
+                    {
+                        executionContext.Error(stderr.Data);
+                    }
+                };
+
+                var trimChars = new[] { '\\', '/' };
+
+                dropLocation = Path.Combine(dropLocation.TrimEnd(trimChars));
+                downloadFolderPath = downloadFolderPath.TrimEnd(trimChars);
+
+                string robocopyArguments = "\"" + dropLocation + "\" \"" + downloadFolderPath + "\" * /E /COPY:DA /NP /R:3";
+
+                int exitCode = await processInvoker.ExecuteAsync(
+                        workingDirectory: "",
+                        fileName: "robocopy",
+                        arguments: robocopyArguments,
+                        environment: null,
+                        requireExitCodeZero: false,
+                        outputEncoding: null,
+                        killProcessOnCancel: true,
+                        cancellationToken: cancellationToken);
+
+                executionContext.Output(StringUtil.Loc("RobocopyBasedPublishArtifactTaskExitCode", exitCode));
+
+                if (exitCode >= 8)
```

Comment:
Make this > 0x07 since value greater than 7 indicates there's at least one failure. 
[https://ss64.com/nt/robocopy-exit.html](https://ss64.com/nt/robocopy-exit.html)

---
In reply to: [318835632](https://github.com/microsoft/azure-pipelines-agent/pull/2406#discussion_r318835632) [](ancestors = 318835632)